dags/utils/dag_utils.py

```diff
@@ -0,0 +1,275 @@
+"""
+Copyright 2023 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License."""
+"""Utility functions for DAGs."""
+
+import ast
+from contextlib import closing
+from dataclasses import asdict
+import datetime
+from functools import partial
+import importlib
+import json
+import os
+import pathlib
+import random
+import sys
+from typing import Any, List, Mapping, Optional, Sequence
+import uuid
+
+from airflow.decorators import dag
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.models import Variable
+from airflow.operators.python_operator import PythonOperator
+import psycopg2
+
+from protocols.destination_proto import DestinationProto
+from protocols.source_proto import SourceProto
+from utils.run_result import RunResult
+
+MAX_TRIES = 3
+
+
+class DagUtils:
+  """A set of utility functions for DAGs."""
+
+  @classmethod
+  def build_dynamic_dag(
+      cls,
+      new_dag_name: str,
+      schedule: str,
+      target_source: SourceProto,
+      target_destination: DestinationProto,
+      dest_type: str,
+      dest_folder: str,
+      dest_config: Mapping[str, Any],
+      reusable_credentials: Optional[Sequence[Any]] = None,
+  ):
+    """Dynamically creates a DAG based on a given connection."""
+    connection_id = f'{new_dag_name}_dag'
+    schedule_interval = schedule if schedule else None
+    start_date = datetime.datetime(2023, 1, 1, 0, 0, 0)
+
+    @dag(dag_id=connection_id,
+         is_paused_upon_creation=False,
+         start_date=start_date,
+         schedule_interval=schedule_interval,
+         catchup=False)
+    def dynamic_generated_dag():
+
+      def process(task_instance, dry_run_str: str) -> None:
+        dry_run = cls.parse_dry_run(connection_id, dry_run_str)
+        fields = target_destination.fields()
+        batch_size = target_destination.batch_size()
+        offset = 0
+        get_data = partial(target_source.get_data,
+                           fields=fields,
+                           limit=batch_size,
+                           reusable_credentials=reusable_credentials)
+        data = get_data(offset=offset)
+
+        run_result = RunResult(0, 0, [], dry_run)
+        while data:
+          run_result += target_destination.send_data(data, dry_run)
+          offset += batch_size
+          ## TESTING ONLY ######################################################
+          # TODO(blevitan): REMOVE this line before merging BLEVITAN_DEV branch into MAIN.
```

Comment:
Remove this line

Reply:
will remove when done testing on rebase.